### PR TITLE
storage: add block load limit and metrics

### PR DESCRIPTION
**Pull request by RaduBerinde**
First commit is just the Pebble bump from #125464.

#### storage: add block load limit and metrics

This change adds a limit on the number of block loads that happen in
parallel, along with per-store metrics for (instantaneous) number of
block loads in progress and the cumulative number of loads that had to
wait because of the limit (i.e. were queued).

Each store has its own limiter; the limit is distributed evenly
between all stores.

Epic: none

### CODE REVIEW

**Comment by RaduBerinde** [comment]
I want to do some additional manual testing, but it seems to work per the unit test I added.

**Review by sean- — APPROVED** [review]

File: pkg/kv/kvserver/metrics.go [hunk 1]
```diff
@@ -809,6 +809,18 @@ bytes preserved during flushes and compactions over the lifetime of the process.
 		Measurement: "Bytes",
 		Unit:        metric.Unit_BYTES,
 	}
+	metaBlockLoadsInProgress = metric.Metadata{
```
Comment (nicktrav) [inline pkg/kv/kvserver/metrics.go:812]: Do we want another metric for time waited? I guess we also have a _total_ block load time. No strong opinion.

**Review by nicktrav — APPROVED** [review]
Nice. LGTM. Perhaps grab a second review from another Storage eng.

File: pkg/kv/kvserver/metrics.go (hunk 1 above)
Comment (RaduBerinde) [reply pkg/kv/kvserver/metrics.go:812]: Not entirely sure what would be useful. The sum of wait times for each load? Or the total amount of time during which there was a non-empty queue? Probably the latter. Either way, that would have to be added to the semaphore so I won't do it in as part of this patch.

**Review by itsbilal — APPROVED** [review]
LGTM

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 3 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-:-O-7Dzyw6IfOucy3s9lQ:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @nicktrav, and @RaduBerinde)

___
*[`pkg/kv/kvserver/metrics.go` line 812 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125468#gh-1634884430:-O-7CyNv8CaHv6fOv4U5:b-jyq3k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b35552666da9753aa2cf6a62ed212c386f2623a7/pkg/kv/kvserver/metrics.go#L812)):*
> Or the total amount of time during which there was a non-empty queue? Probably the latter. Either way, that would have to be added to the semaphore so I won't do it in as part of this patch.

+1
___
*[`pkg/storage/pebble.go` line 979 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-O-7DKt-9s-5t4DmW-FN:-O-7DKt-9s-5t4DmW-FO:bbl7cau) ([raw file](https://github.com/cockroachdb/cockroach/blob/b35552666da9753aa2cf6a62ed212c386f2623a7/pkg/storage/pebble.go#L979)):*
> ```Go
> 	// concurrency limit: it is the current valuer of the
> 	// BlockLoadConcurrencyLimit setting divided by this value. It should be set
> 	// to the number of stores.
> ```

Why are we dividing instead of sharing the same semaphore across stores?
Former seems like unnecessary "resource fragmentation".

I suppose the concern is that a bad disk on one store can cause it to consume all the slots and not release them. Is worth a code comment.

**Review by RaduBerinde — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-:-O-7HhSC7ue8XWqU65Pf:b-814g8k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @nicktrav, and @sumeerbhola)

___
*[`pkg/storage/pebble.go` line 979 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-O-7DKt-9s-5t4DmW-FN:-O-7HHsV9PuV6FXs9vJe:bts6r9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b35552666da9753aa2cf6a62ed212c386f2623a7/pkg/storage/pebble.go#L979)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Why are we dividing instead of sharing the same semaphore across stores?
Former seems like unnecessary "resource fragmentation".

I suppose the concern is that a bad disk on one store can cause it to consume all the slots and not release them. Is worth a code comment.
</blockquote></details>

Correct, I added some comments. It's also nice that the stores don't have to synchronize with each other, and that we get separate metrics per store.

**Review by jbowens — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 4 files at r1, 5 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-:-O-7Sown18tcdAh0Jicn:bg7igkv)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @nicktrav and @sumeerbhola)

**Review by sean- — APPROVED** [review]
> Reviewed 12 of 12 files at r3, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125468)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @nicktrav and @sumeerbhola)

___
*[`pkg/kv/kvserver/metrics.go` line 2627 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-O-COM2T650W1DWObjOk:-O-COM2T650W1DWObjOl:b-4kl023) ([raw file](https://github.com/cockroachdb/cockroach/blob/788d9253105342bcb9e087778d26f4bd51c83dd2/pkg/kv/kvserver/metrics.go#L2627)):*
> ```Go
> 	SharedStorageBytesWritten         *metric.Gauge
> 	BlockLoadsInProgress              *metric.Gauge
> 	BlockLoadsQueued                  *metric.Gauge
> ```

Do we want to bite the bullet here and have this metric be a counter?  I get that everything here is cargo-culted.
___
*[`pkg/storage/pebble.go` line 979 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-O-7DKt-9s-5t4DmW-FN:-O-CNDqOBpMFoSBEuTEo:b-krulhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b35552666da9753aa2cf6a62ed212c386f2623a7/pkg/storage/pebble.go#L979)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Correct, I added some comments. It's also nice that the stores don't have to synchronize with each other, and that we get separate metrics per store.
</blockquote></details>

Because if a single store goes sideways for some reason, it shouldn't impact the remaining healthy stores.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125468)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @nicktrav and @sumeerbhola)

___
*[`pkg/kv/kvserver/metrics.go` line 2627 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125468#-O-COM2T650W1DWObjOk:-O-COzIj6aZNCCe-Xp22:b-s0whuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/788d9253105342bcb9e087778d26f4bd51c83dd2/pkg/kv/kvserver/metrics.go#L2627)):*
<details><summary><i>Previously, sean- (Sean Chittenden) wrote…</i></summary><blockquote>

Do we want to bite the bullet here and have this metric be a counter?  I get that everything here is cargo-culted.
</blockquote></details>

I can't use a Counter because it cannot be set to a specific value, just incremented/decremented. This is the reason most things here are gauges. There's a comment above. Definitely won't tackle this for something that's getting backported.

**Review by sean- — COMMENTED** [review]
> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125468)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale)

**Pull request merged**